src/core/utils/conversation/helpers.ts

```diff
@@ -189,3 +192,37 @@ const removeInitialSystemMessage = (conversation: ConversationHistory) => {
     ? conversation.conversation.slice(1)
     : conversation.conversation;
 };
+
+//  todo - put on model configuration
+export const MAX_TOKENS = 128000;
+
+/**
+ * Pure function to calculate context metrics based on messages and max tokens
+ * @param messages Array of chat messages
+ * @returns ContextMetrics object
+ */
+export async function calculateContextMetrics(
+  chatMessages: ChatMessage[],
+): Promise<ContextMetrics> {
+  const messages = chatMessages as TextChatMessage[];
+  const maxTokens = MAX_TOKENS;
+  const tokensUsed = encodeChat(messages, 'gpt-4o').length;
```

Comment:
We can encode the entire chat at once, so long as its one of the string literals it expects.